CLC documentation [API-1340] (#462)
hazelcast/hz-docs: docs/modules/clients/pages/clc.adoc

> = Hazelcast Command-Line Client

Reviewer: How about "Command Line Client" instead of "Hazelcast Command-Line Client" ?

Reviewer 2: I think the existing title is good. Users will likely search Google for the Hazelcast CLC so having Hazelcast in the title is helpful.

:page-api-reference: https://github.com/hazelcast/hazelcast-commandline-client
:page-toclevels: 3

> [[clc-client]]

Reviewer 2: ```suggestion
:page-beta: true
```

This will add a 'beta' label to the title at the top of the page.

Author: Great, added it https://github.com/hazelcast/hz-docs/pull/462/commits/f8ae9313c19cc809466021a2dae1a074fcb090f2

> Hazelcast Command-Line Client or *hzc* in short, is a command-line tool.

Reviewer: I would change it to: "Hazelcast Command Line Client or CLC ..."

Reviewer 2: Are we calling it hzc for short? Shouldn't we call it Hazelcast CLC? Since this tool also includes commands for Hazelcast Cloud, which uses hzcloud.

I'd recommend something like this for the overview:

> The Hazelcast Command-Line Client (CLC) is a cross-platform command-line tool to connect to Hazelcast Platform and Hazelcast Cloud. You can use the CLI to execute commands through a terminal, using interactive command-line prompts, or a script.

> You can install the Hazelcast CLC locally on Linux or Mac computers.

Author: @JakeSCahill your recommendation looks great. I think we should drop cross-platform for now since Windows is not supported yet.

I think it also misses the actual description of the tool. We want users to mainly use it to operate on distributed data structures. There is also just few cluster utility commands. I think we need to describe this too somehow. Wdyt?

Author: Refactored Jake's suggestion a bit. Let me know if it doesn't work, please. https://github.com/hazelcast/hz-docs/pull/462/commits/b3e94f20a0864ac722d6b633d8dc5a57e9758e7b

It provides access to Hazelcast Cluster similar to Hazelcast Clients (Java, C++, .NET, Nodejs, Python, Go) with commands as opposed to writing actual code.

> == Installation

Author: https://github.com/hazelcast/hz-docs/pull/462/commits/b3e94f20a0864ac722d6b633d8dc5a57e9758e7b

> To get started, you need to install the *hzc* (Hazelcast Command-line Client). There are two methods to install.

Author: Thanks, changed it a bit https://github.com/hazelcast/hz-docs/pull/462/commits/b3e94f20a0864ac722d6b633d8dc5a57e9758e7b

> You can install it with a custom script

Author: Great tip! https://github.com/hazelcast/hz-docs/pull/462/commits/b3e94f20a0864ac722d6b633d8dc5a57e9758e7b

[source,bash,subs="attributes+"]
----
curl https://raw.githubusercontent.com/hazelcast/hazelcast-commandline-client/main/scripts/install.sh | bash
----
or as a brew package
[source,bash,subs="attributes+"]
----
brew tap hazelcast/homebrew-hz
brew install hazelcast-commandline-client
----
NOTE: To take advantage of auto-completion support, follow the instructions that appear during installation.

> == Uninstallation

Author: Changed as suggested https://github.com/hazelcast/hz-docs/pull/462/commits/b3e94f20a0864ac722d6b633d8dc5a57e9758e7b

You can choose the corresponding uninstallation option.

- Uninstall the brew package
[source,bash,subs="attributes+"]
----
brew uninstall hazelcast-commandline-client
brew untap hazelcast/homebrew-hz            # optional
----
- Uninstall the script installation
[source,bash,subs="attributes+"]
----
~/.local/share/hz-cli/bin/uninstall.sh
----

> == Usage

Author: Not needed, you are right. https://github.com/hazelcast/hz-docs/pull/462/commits/b3e94f20a0864ac722d6b633d8dc5a57e9758e7b

> == Configuring HZC

Author: https://github.com/hazelcast/hz-docs/pull/462/commits/b3e94f20a0864ac722d6b633d8dc5a57e9758e7b

> Hzc uses xref:go.adoc[hazelcast-go-client] under the hood to connect to cluster and to make operations.

Author: Thanks for the rewording, changed as suggested https://github.com/hazelcast/hz-docs/pull/462/commits/b3e94f20a0864ac722d6b633d8dc5a57e9758e7b

=== Configuring SSL

You can use the following configuration file to enable SSL support:

[source,yaml]
----
ssl:
  servername: "HOSTNAME-FOR-SERVER"
  # or: insecureskipverify: true
hazelcast:
  cluster:
    security:
      credentials:
        username: "OPTIONAL USERNAME"
        password: "OPTIONAL PASSWORD"
    name: "CLUSTER-NAME"
    network:
    addresses:
    - "localhost:5701"
----

Mutual authentication is also supported:
[source,yaml]
----
ssl:
    servername: "HOSTNAME-FOR-SERVER"
    # insecureskipverify: true
    capath: "/tmp/ca.pem"
    certpath: "/tmp/cert.pem"
    keypath: "/tmp/key.pem"
    keypassword: "PASSWORD FOR THE KEY"
hazelcast:
  cluster:
    security:
      credentials:
        username: "OPTIONAL USERNAME"
        password: "OPTIONAL PASSWORD"
    name: "CLUSTER-NAME"
    network:
      addresses:
        - "localhost:5701"
----

Cloud SSL configuration:

[source,yaml]
----
ssl:
    capath: "/tmp/ca.pem"
    certpath: "/tmp/cert.pem"
    keypath: "/tmp/key.pem"
    keypassword: "PASSWORD FOR THE KEY"
hazelcast:
  cluster:
    name: "CLUSTER NAME"
    cloud:
      token: "HAZELCAST CLOUD TOKEN"
      enabled: true
----

> === Modes

Author: Makes sense, https://github.com/hazelcast/hz-docs/pull/462/commits/b3e94f20a0864ac722d6b633d8dc5a57e9758e7b

> *hzc* packs 3 usage modes to simplify different use cases.

Reviewer 2: ```suggestion
You can use the Hazelcast CLC in three modes:

- Scripting
- Interactive
- SQL
```

Author: Thanks, changed https://github.com/hazelcast/hz-docs/pull/462/commits/b3e94f20a0864ac722d6b633d8dc5a57e9758e7b

> *Default (Non-interactive)*

Author: Changed as suggested https://github.com/hazelcast/hz-docs/pull/462/commits/b3e94f20a0864ac722d6b633d8dc5a57e9758e7b

You can execute commands similar to any other command line tool. It can also be useful for scripting.
[source,bash,subs="attributes+"]
----
hzc map --cluster-name cluster1 --name myMap put --key myKey --value myValue
----

> *Interactive*

Reviewer: I guess this should be a heading? (similar to other modes)

Author: I couldn't heading option with that hierarchy depth. Maybe @JakeSCahill can give an opinion on how to organize them.

Author: By eliminating one of the header levels, I was able to convert these to headings, as you suggested https://github.com/hazelcast/hz-docs/pull/462/commits/b3e94f20a0864ac722d6b633d8dc5a57e9758e7b

> *Interactive*

Author: Changed as suggested https://github.com/hazelcast/hz-docs/pull/462/commits/b3e94f20a0864ac722d6b633d8dc5a57e9758e7b

This starts an interactive prompt, in which you can interactively navigate commands and explore possible subcommands and flags. It is also useful for manual tasks.

image:ROOT:hzc-interactive-screenshot.png[HZC interactive]

[source,bash,subs="attributes+"]
----
hzc
----

> On this mode, you can also leverage the "use" command to avoid re-typing the distributed object name on each command. Note that map name "m1" is referenced only once.

Reviewer: "In this mode" ?

Author: Thanks, fixed https://github.com/hazelcast/hz-docs/pull/462/commits/b3e94f20a0864ac722d6b633d8dc5a57e9758e7b

----
hzc
Connecting to the cluster ...
hzc localhost:5701@dev> map use m1
hzc localhost:5701@dev&m:m1> map put -k k1 -v v1
hzc localhost:5701@dev&m:m1> map get -k k1
v1
hzc localhost:5701@dev&m:m1>
----

> *SQL Browser*

Reviewer 2: ```suggestion
=== SQL Mode
```

Author: https://github.com/hazelcast/hz-docs/pull/462/commits/b3e94f20a0864ac722d6b633d8dc5a57e9758e7b

All the modes support executing SQL statements. In addition to that, this mode makes it easier to navigate results with an interactive browser.

image:ROOT:hzc-sql-browser-screenshot.png[HZC SQL Browser]]

To start the browser simply type

[source,bash,subs="attributes+"]
----
hzc sql
----

> == Commands

Author: I think we should have a general section that groups all commands. I understand that this does not provide any information hence it is against the styling guides. But I believe there is value in this type of grouping. WDYT?

> === Distributed Data Structures

Reviewer 2: ```suggestion
```

I don't think we need this. It doesn't add any value and we don't introduce the section with any additional details.

Author: Makes sense thanks https://github.com/hazelcast/hz-docs/pull/462/commits/b3e94f20a0864ac722d6b633d8dc5a57e9758e7b

> ==== Map

Reviewer 2: ```suggestion
== hzc map
```

Author: I liked this idea. Changed it as you suggested. Tagging @yuce to also highlight this for him. https://github.com/hazelcast/hz-docs/pull/462/commits/b3e94f20a0864ac722d6b633d8dc5a57e9758e7b

** mutually exclusive with "--value", "--value-file", "--value-type"

[source,bash,subs="attributes+"]
----
# Keys and values are matched with the given order
hzc map put-all -n mapname --key-type int16 -k 1 -k 2 --value-type json -f valueFile.json -v '{"field":"tmp"}'
hzc map put-all -n mapname --json-entry entries.json`
----

Example json-entry file:
```
{
     "key1": "value1",
     "key2": {
       "innerData": "data",
       "anotherInnerData": 5.0
     },
     "key3": true,
     "key4": [1, 2, 3, 4, 5]
   }
```

*remove*: Deletes the value for the given key from the specified map.

* --name/-n: name of the map
* --key/-k: key of the entry
* [optional] --key-type: specifies the type of the key
**   possible values: string [default], boolean, json, int8, int16, int32, int64, float32, float64

` hzc map remove -n mapname -k k1`

*use*: Sets the default map name (interactive-mode only).

* --reset: unset default name for map

[source,bash,subs="attributes+"]
----
Available Commands:
hzc map use m1    # sets the default map name to m1 unless set explicitly
hzc map get --key k1    # "--name m1" is inferred
hzc map use --reset	  # resets the behaviour
----

> ==== Cluster

Author: Thanks, changed as suggested https://github.com/hazelcast/hz-docs/pull/462/commits/b3e94f20a0864ac722d6b633d8dc5a57e9758e7b

Cluster commands are a group of administrative cluster operation.

> Hzc uses REST-API for cluster management operations. In order to perform them, you have to enable the REST-API in the "member"

Reviewer 2: ```suggestion
You must enable the REST API on all members of your cluster to use the cluster commands. These commands use the Hazelcast REST API for cluster management operations. To enable REST API, follow the instructions in the member log or see the documentation: xref:maintain-cluster:rest-api.adoc#enabling-rest-api[enabling-rest-api].
```

Paragraphs should follow the what, why, how structure.

Author: Thanks for the tip! https://github.com/hazelcast/hz-docs/pull/462/commits/b3e94f20a0864ac722d6b633d8dc5a57e9758e7b

> configuration (which is disabled by default).

Author: https://github.com/hazelcast/hz-docs/pull/462/commits/b3e94f20a0864ac722d6b633d8dc5a57e9758e7b

> To enable REST API, follow the instructions in the member log or see the documentation: xref:maintain-cluster:rest-api.adoc#enabling-rest-api[enabling-rest-api]

Reviewer 2: ```suggestion
```

Author: https://github.com/hazelcast/hz-docs/pull/462/commits/b3e94f20a0864ac722d6b633d8dc5a57e9758e7b

*change-state*: Changes the state of the cluster.

* --state/-s: new state of the cluster
** possible values: active, no_migration, frozen, passive

`hzc cluster change-state --state passive`

*get-state*: Gets the state of the cluster.

`hzc cluster get-state`

*shutdown*: Shutdowns the cluster

`hzc cluster shutdown`

*version*: Retrieves and prints the version of the cluster

`hzc cluster version`

> ==== SQL

Reviewer 2: ```suggestion
== hzc sql
```

Author: https://github.com/hazelcast/hz-docs/pull/462/commits/b3e94f20a0864ac722d6b633d8dc5a57e9758e7b

You can execute a SQL statement or start the SQL Browser with the "sql" command.

`hzc sql` starts the SQL Browser.

`hzc sql "select * from employees"` executes the given query. You can also change the output format with "--output" or "-o" flags. Supported styles are "csv" and "pretty"[default].

> === Querying with SQL

Reviewer 2: ```suggestion
== Querying with SQL
```

Author: Changed this as suggested https://github.com/hazelcast/hz-docs/pull/462/commits/b3e94f20a0864ac722d6b633d8dc5a57e9758e7b

> You can query a map with SQL using 3 different modes of *hzc*.

Reviewer 2: ```suggestion
You can use SQL to query a map, using all three modes of the Hazelcast CLC.
```

Author: Much better! https://github.com/hazelcast/hz-docs/pull/462/commits/b3e94f20a0864ac722d6b633d8dc5a57e9758e7b

> Default:

Reviewer 2: It's better to put this content in tabs. Make sure to use the same mode names that are used in the 'Modes' section

Author: Done! https://github.com/hazelcast/hz-docs/pull/462/commits/b3e94f20a0864ac722d6b633d8dc5a57e9758e7b

= Hazelcast Command-Line Client
:page-api-reference: https://github.com/hazelcast/hazelcast-commandline-client
:page-toclevels: 3
:page-beta: true

> The Hazelcast Command-Line Client (CLC) is a command-line tool to connect and operate on Hazelcast Platform and Hazelcast Cloud. You can use the CLC to execute commands through a terminal.

Reviewer 2: ```suggestion
The Hazelcast Command-Line Client (CLC) is a command-line tool to connect to and operate on Hazelcast Platform and Hazelcast Cloud. You can use the CLC to execute commands through a terminal.
```
